godot-core/src/builtin/basis.rs

```diff
@@ -0,0 +1,980 @@
+        let [Vector3 {
+            x: x0,
+            y: y0,
+            z: z0,
+        }, Vector3 {
+            x: x1,
+            y: y1,
+            z: z1,
+        }, Vector3 {
+            x: x2,
+            y: y2,
+            z: z2,
+        }] = self.rows;
+        [x0, y0, z0, x1, y1, z1, x2, y2, z2]
+    }
+
+    /// Returns the scale of the matrix.
+    ///
+    /// _Godot equivalent: `Basis.get_scale()`_
+    #[must_use]
+    pub fn scale(&self) -> Vector3 {
+        let det = self.determinant();
+        let det_sign = if det < 0.0 { -1.0 } else { 1.0 };
+
+        Vector3::new(
+            self.col_x().length(),
+            self.col_y().length(),
+            self.col_z().length(),
+        ) * det_sign
+    }
+
+    /// Returns the rotation of the matrix in euler angles.
+    ///
+    /// The order of the angles are given by `order`.
+    ///
+    /// _Godot equivalent: `Basis.get_euler()`_
+    #[must_use]
+    pub fn euler_angles(&self, order: EulerOrder) -> Vector3 {
+        // Translated from "Basis::get_euler" in
+        // https://github.com/godotengine/godot/blob/master/core/math/basis.cpp
+
+        // We can't use glam to do these conversions since glam uses intrinsic rotations
+        // whereas godot uses extrinsic rotations.
+        // see https://github.com/bitshifter/glam-rs/issues/337
+        match order {
+            EulerOrder::XYZ => {
+                // Euler angles in XYZ convention.
+                // See https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix
+                //
+                // rot =  cy*cz          -cy*sz           sy
+                //        cz*sx*sy+cx*sz  cx*cz-sx*sy*sz -cy*sx
+                //       -cx*cz*sy+sx*sz  cz*sx+cx*sy*sz  cx*cy
+
+                let sy: f32 = self[0].z;
+                if sy < 1.0 - CMP_EPSILON {
+                    if sy > -(1.0 - CMP_EPSILON) {
+                        // is this a pure Y rotation?
+                        if self[1].x == 0.0
+                            && self[0].y == 0.0
+                            && self[1].z == 0.0
+                            && self[2].y == 0.0
+                            && self[1].y == 1.0
+                        {
+                            // return the simplest form (human friendlier in editor and scripts)
+                            Vector3::new(0.0, f32::atan2(self[0].z, self[0].x), 0.0)
+                        } else {
+                            Vector3::new(
+                                f32::atan2(-self[1].z, self[2].z),
+                                f32::asin(sy),
+                                f32::atan2(-self[0].y, self[0].x),
+                            )
+                        }
+                    } else {
+                        Vector3::new(f32::atan2(self[2].y, self[1].y), -FRAC_PI_2, 0.0)
+                    }
+                } else {
+                    Vector3::new(f32::atan2(self[2].y, self[1].y), FRAC_PI_2, 0.0)
+                }
```

Comment:
This exists 6 times in a slight variantion. Is it not possible to extract the core operations to a function, by reordering elements before and after the operation? In principle:
```rs
use glam::swizzles::Vec3Swizzles as _;

// Get input in right order
let input: Vector3 = ...;
let zxy: glam::Vec3 = input.to_glam().zxy();

// Actual operation
let zxy: glam::Vec3 = transmogrify(zxy);

// Transform back
let xyz: Vector3 = zxy.yzx().to_front();
```

Might need more than 1 vector, but you get the idea.

Reply:
i'll look into it

Reply:
ok i was able to do something close, but not exact. There are like 3 edge cases that make it not work perfectly. but at least for 4/6 of the euler orderings it's basically just that.